### PR TITLE
feat: Add kata-containers extension

### PR DESCRIPTION
**Pull request by fidencio**
Kata Containers provides an OCI runtime that focuses on protecting the host from malicious workloads, taking advantage of KVM to provide an extra isolation layer.

Kata Containers is also the foundation piece for Confidential Containers, as it's the most suitable OCI runtime to be used with Trusted Execution Environments.

Having Kata Containers here, even restricting it to be used with only one of its drivers (for now), opens the path for future collaboration, and providing Talos a reasonable path to become a Kubernetes distro that's TEE capable.

For now we're sticking to using Cloud Hypervisor as the preferred driver for Kata Containers, which probably could change in the future, but we don't want to start with a situation where we'll increase the image size by a whole lot, thus taking the smallest footprint that can be achieved based on Kata Containers stable releases.

Depends on: https://github.com/siderolabs/talos/pull/8287

Kata Containers: https://katacontainers.io/
Cloud Hypervisor: https://www.cloudhypervisor.org/
Confidential Containers: https://github.com/confidential-containers

### CODE REVIEW

File: container-runtime/kata-containers/vars.yaml [hunk 1]
```diff
@@ -0,0 +1 @@
+VERSION: "{{ .GVISOR_VERSION }}-{{ .BUILD_ARG_TAG }}"
```
Comment (frezbo) [inline container-runtime/kata-containers/vars.yaml:1]:
```suggestion
VERSION: "{{ .KATA_CONTAINERS_VERSION }}"
```

The `BUILD_ARG_TAG` is unnecessary if the runtime doesn't depend on kernel modules/ something that's tied to specific talos release

**Comment by frezbo** [comment]
If you're interested in adding tests can take a look at https://github.com/siderolabs/talos/blob/main/internal/integration/api/extensions_qemu.go#L365

We could help you in testing, otherwise don't sweat it, I can add a test.

File: Makefile [hunk 2]
```diff
@@ -66,6 +66,7 @@ TARGETS += i915-ucode
 TARGETS += intel-ice-firmware
 TARGETS += intel-ucode
 TARGETS += iscsi-tools
+TARGETS += kata-containers
```
[the remark this answers is not in the document]
Comment (fidencio) [reply Makefile:69]:
```
⋊> extensions on main ⨯ make rekres
latest: Pulling from siderolabs/kres
Digest: sha256:36a8b5e8947500373f12ae9149be7cab2186261d3bfa194171271c5a8944bcca
Status: Image is up to date for ghcr.io/siderolabs/kres:latest
ghcr.io/siderolabs/kres:latest
gen started
GITHUB_TOKEN is missing, GitHub API integration is disabled
Error: failed to parse remote URL: origin	https://github.com/siderolabs/extensions (fetch)
origin	https://github.com/siderolabs/extensions (push)
Usage:
  kres gen [flags]

Flags:
  -h, --help   help for gen

failed to parse remote URL: origin	https://github.com/siderolabs/extensions (fetch)
origin	https://github.com/siderolabs/extensions (push)
make: *** [Makefile:190: rekres] Error 1
```

Not exactly sure what I'm missing here.

**Comment by frezbo** [comment]
> > We could help you in testing, otherwise don't sweat it, I can add a test.
> 
> Adding a test should be quite straightforward for what I can see. I can add one. The main question is, shall the test be added after this one gets merge? Is there a way to have the test added on a different repo depending on this PR here?

Currently the test can run in CI only after this gets merged, usually I test it locally, then approve the extensions PR

File: container-runtime/kata-containers/pkg.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  - sources:
+      - url: https://github.com/kata-containers/kata-containers/releases/download/{{ .VERSION }}/kata-static-{{ .VERSION }}-amd64.tar.xz
+        destination: kata-static.tar.xz
+    prepare:
+      - |
+        sed -i 's#$VERSION#{{ .VERSION }}#' /pkg/manifest.yaml
+      - |
+        mkdir -p /tmp/kata-containers
+
+        tar xvJpf  ata-static.tar.xz -C /tmp/kata-containers/
```
Comment (sergelogvinov) [inline container-runtime/kata-containers/pkg.yaml:16]: Typo: ata-static.tar.xz -> kata-static.tar.xz

**Comment by frezbo** [comment]
Okay, seems we cannot use the static archives from kata, `containerd-shim-kata-v2` seems to be linked against glibc

**Comment by frezbo** [comment]
> Okay, seems we cannot use the static archives from kata, `containerd-shim-kata-v2` seems to be linked against glibc

Okay, i got around this by using the rust runtime which is statically compiled, but not getting this error:

```text
  Warning  FailedCreatePodSandBox  10s   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: Others("failed to handler message try init runtime instance\n\nCaused by:\n    0: load config\n    1: load toml config\n    2: No such file or directory (os error 2)"): unknown
```

**Comment by frezbo** [comment]
> > Okay, seems we cannot use the static archives from kata, `containerd-shim-kata-v2` seems to be linked against glibc
> 
> Okay, i got around this by using the rust runtime which is statically compiled, but not getting this error:
> 
> ```
>   Warning  FailedCreatePodSandBox  10s   kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: Others("failed to handler message try init runtime instance\n\nCaused by:\n    0: load config\n    1: load toml config\n    2: No such file or directory (os error 2)"): unknown
> ```

Fixed the config file path and now getting this error:

```text
  Warning  FailedCreatePodSandBox  2s    kubelet            Failed to create pod sandbox: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: Others("failed to handler message try init runtime instance\n\nCaused by:\n    0: load config\n    1: load toml config\n    2: Can not find plugin for hypervisor clh"): unknown
```

I'm not sure why, the `hypervisor.clh` section correctly defined the path

```bash
talosctl -n 10.5.0.3 read /usr/local/share/kata-containers/configuration.toml | head -n 15

[hypervisor.clh]
path = "/usr/local/bin/cloud-hypervisor"
kernel = "/usr/local/share/kata-containers/vmlinux.container"
initrd = "/usr/local/share/kata-containers/kata-containers-initrd.img"

# rootfs filesystem type:
#   - ext4 (default)
#   - xfs
#   - erofs
rootfs_type="ext4"

# Enable confidential guest support.
# Toggling that setting may trigger different hardware features, ranging
# from memory encryption to both memory and CPU-state encryption and integrity.
# The Kata Containers runtime dynamically detects the available feature set and

```

**Comment by fidencio** [comment]
> Okay, seems we cannot use the static archives from kata, `containerd-shim-kata-v2` seems to be linked against glibc

It's statically linked against glibc, is this a problem?
We can easily switch to providing a binary that's linked against musl, I think.

**Comment by fidencio** [comment]
The rust runtime is a no-go yet, as that's not on a stable release and not passing all our tests.
We plan to officially switch to the rust runtime as part of 4.0.0, but for now we should stick to the golang one, even if it means providing a musl linked runtime.

**Comment by frezbo** [comment]
More updates:

* building the go runtime fixed the ` Can not find plugin for hypervisor clh` error, but got a syslog error
* adding a dummy `syslog` extension fixed the issue and able to schedule pod.

**Review by frezbo — APPROVED** [review]
Thanks you :heart: , now that https://github.com/siderolabs/talos/pull/8287 is part of Talos, this can be merged.

**Pull request merged**